File: python/gripsack/fetch.py
"""Typed fetchers (0001 §3.1, 0002 §2)."""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Any, Optional


class FetchKind(str, Enum):
    """The closed set of fetch kinds the engine interprets."""

    GITHUB_RELEASE = "github_release"
    TARBALL = "tarball"
    GIT = "git"
    FILE = "file"
    PLUGIN = "plugin"
    BREW = "brew"
    PIXI = "pixi"


@dataclass(frozen=True)
class Fetch:
    """A fetch spec — the engine interprets it; never a script."""

    kind: FetchKind
    args: dict[str, Any]

    def __post_init__(self) -> None:
        object.__setattr__(self, "kind", FetchKind(self.kind))

    def to_ir(self) -> dict[str, Any]:
        return {"kind": self.kind.value, **self.args}


def github_release(
    repo: str,
    asset: str,
    version: Optional[str] = None,
    sha256: Optional[str] = None,
    base_url: Optional[str] = None,
) -> Fetch:
    """GitHub releases; `base_url` covers GitHub Enterprise (0002 rung 1)."""
    args: dict[str, Any] = {"repo": repo, "asset": asset}
    if version is not None:
        args["version"] = version
    if sha256 is not None:
        args["sha256"] = sha256
    if base_url is not None:
        args["base_url"] = base_url
    return Fetch(FetchKind.GITHUB_RELEASE, args)


def tarball(url: str, sha256: Optional[str] = None) -> Fetch:
    args: dict[str, Any] = {"url": url}
    if sha256 is not None:
        args["sha256"] = sha256
    return Fetch(FetchKind.TARBALL, args)


def git(url: str, rev: str) -> Fetch:
    return Fetch(FetchKind.GIT, {"url": url, "rev": rev})


def file_fetch(path: str) -> Fetch:
    return Fetch(FetchKind.FILE, {"path": path})


def plugin_fetch(name: str, **args: Any) -> Fetch:
    """A fetcher plugin transport (0002 §4) — `gripfetch-<name>`."""
    return Fetch(FetchKind.PLUGIN, {"name": name, "args": args})


def brew(formula: str, version: Optional[str] = None) -> Fetch:
    """A Homebrew bottle — resolved from the formula JSON, so the pin
    (bottle sha256) needs no download at update time.

    `version` is a tripwire, not a range: brew only ever serves the
    *current* formula, so a mismatch fails at resolve with
    "`grip update` to move" rather than a sha mismatch later."""
    args: dict[str, Any] = {"formula": formula}
    if version:
        args["version"] = version
    return Fetch(FetchKind.BREW, args)


def pixi(package: str, version: Optional[str] = None) -> Fetch:
    """A conda package via pixi, installed into an isolated PIXI_HOME
    and harvested into the store."""
    args: dict[str, Any] = {"package": package}
    if version is not None:
        args["version"] = version
    return Fetch(FetchKind.PIXI, args)

File: typescript/src/fetch.ts
/** Typed fetchers (0001 §3.1, 0002 §2). */

export type Fetch =
  | { kind: "github_release"; repo: string; asset: string; version?: string; sha256?: string; base_url?: string }
  | { kind: "tarball"; url: string; sha256?: string }
  | { kind: "git"; url: string; rev?: string }
  | { kind: "file"; path: string }
  | { kind: "plugin"; name: string; args?: Record<string, unknown> }
  | { kind: "brew"; formula: string; version?: string; sha256?: string }
  | { kind: "pixi"; package: string; version?: string; sha256?: string };

/**
 * GitHub release fetcher. `asset` patterns accept `{version}` (the
 * tag, either v-form) and the platform placeholders (0016 §D1):
 * `{system}` (x86_64-linux, flake-style), `{target}` (the rust
 * triple, musl for linux), `{arch}` (x86_64|aarch64), `{arch.go}`
 * (amd64|arm64), `{os}` (linux|darwin) — expanded by the core from the
 * machine's facts, so one spec serves every platform.
 */
export function githubRelease(spec: {
  repo: string;
  asset: string;
  version?: string;
  sha256?: string;
  base_url?: string;
}): Fetch {
  return { kind: "github_release", ...spec };
}

/** Direct tarball/zip URL; accepts the same platform placeholders as
 *  githubRelease (0016 §D1). */
export function tarball(url: string, sha256?: string): Fetch {
  return sha256 === undefined
    ? { kind: "tarball", url }
    : { kind: "tarball", url, sha256 };
}

export function git(url: string, rev?: string): Fetch {
  // rev absent = float: the core pins the default branch's HEAD into
  // the lockfile at resolve time; `grip update` moves it (0016 §D2)
  return rev === undefined
    ? { kind: "git", url }
    : { kind: "git", url, rev };
}

export function fileFetch(path: string): Fetch {
  return { kind: "file", path };
}

/** A fetcher plugin transport (0002 §4) — `gripfetch-<name>`. */
export function pluginFetch(name: string, args?: Record<string, unknown>): Fetch {
  return args === undefined
    ? { kind: "plugin", name }
    : { kind: "plugin", name, args };
}

/** A Homebrew bottle — resolved from the formula JSON, so the pin
 *  needs no download at update time. `version` is a tripwire: a
 * mismatch fails at resolve (`grip update` to move), never a range. */
export function brew(formula: string, version?: string): Fetch {
  return version === undefined
    ? { kind: "brew", formula }
    : { kind: "brew", formula, version };
}

/** A conda package via pixi, isolated PIXI_HOME, harvested to store. */
export function pixi(pkg: string, version?: string): Fetch {
  return version === undefined
    ? { kind: "pixi", package: pkg }
    : { kind: "pixi", package: pkg, version };
}
